## Загрузка данных

In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры

db_config = {'user': '', # имя пользователя
'pwd': '', # пароль
'host': '',
'port': , # порт подключения
'db': ''} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Базу данных подключили, можем приступать к заданию таблиц.

In [ ]:
for data in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = (
        '''
            select * 
            from {} 
    
        '''.format(data))

    data = pd.io.sql.read_sql(query, con = engine)

Приступим к исследованию таблиц.

## Исследование таблиц

In [8]:
# создадим функцию для вывода общей информации: это позволит избавиться от повторов и сократить код 

def background(data):
    display(data.columns, data.head(), data.describe())
    print('*' * 50)
    data.info()
    print('*' * 50)



In [9]:
for data in [books, authors, publishers, ratings, reviews]:
    background(data)

Index(['book_id', 'author_id', 'title', 'num_pages', 'publication_date',
       'publisher_id'],
      dtype='object')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

book_id    author_id   num_pages  publisher_id
count  1000.000000  1000.000000  1000.00000   1000.000000
mean    500.500000   320.417000   389.11100    171.270000
std     288.819436   181.620172   229.39014     99.082685
min       1.000000     1.000000    14.00000      1.000000
25%     250.750000   162.750000   249.00000     83.000000
50%     500.500000   316.500000   352.00000    177.500000
75%     750.250000   481.000000   453.00000    258.000000
max    1000.000000   636.000000  2690.00000    340.000000

**************************************************

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 1000 entries, 0 to 999

Data columns (total 6 columns):

 #   Column            Non-Null Count  Dtype 

---  ------            --------------  ----- 

 0   book_id           1000 non-null   int64 

 1   author_id         1000 non-null   int64 

 2   title             1000 non-null   object

 3   num_pages         1000 non-null   int64 

 4   publication_date  1000 non-null   object

 5   publisher_id      1000 non-null   int64 

dtypes: int64(4), object(2)

memory usage: 47.0+ KB

**************************************************


Index(['author_id', 'author'], dtype='object')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

author_id
count  636.000000
mean   318.500000
std    183.741666
min      1.000000
25%    159.750000
50%    318.500000
75%    477.250000
max    636.000000

**************************************************

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 636 entries, 0 to 635

Data columns (total 2 columns):

 #   Column     Non-Null Count  Dtype 

---  ------     --------------  ----- 

 0   author_id  636 non-null    int64 

 1   author     636 non-null    object

dtypes: int64(1), object(1)

memory usage: 10.1+ KB

**************************************************


Index(['publisher_id', 'publisher'], dtype='object')

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publisher_id
count    340.000000
mean     170.500000
std       98.293777
min        1.000000
25%       85.750000
50%      170.500000
75%      255.250000
max      340.000000

**************************************************

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 340 entries, 0 to 339

Data columns (total 2 columns):

 #   Column        Non-Null Count  Dtype 

---  ------        --------------  ----- 

 0   publisher_id  340 non-null    int64 

 1   publisher     340 non-null    object

dtypes: int64(1), object(1)

memory usage: 5.4+ KB

**************************************************


Index(['rating_id', 'book_id', 'username', 'rating'], dtype='object')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

rating_id      book_id       rating
count  6456.000000  6456.000000  6456.000000
mean   3228.500000   510.574195     3.928284
std    1863.831001   284.141636     0.943303
min       1.000000     1.000000     1.000000
25%    1614.750000   291.000000     3.000000
50%    3228.500000   506.000000     4.000000
75%    4842.250000   750.000000     5.000000
max    6456.000000  1000.000000     5.000000

**************************************************

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 6456 entries, 0 to 6455

Data columns (total 4 columns):

 #   Column     Non-Null Count  Dtype 

---  ------     --------------  ----- 

 0   rating_id  6456 non-null   int64 

 1   book_id    6456 non-null   int64 

 2   username   6456 non-null   object

 3   rating     6456 non-null   int64 

dtypes: int64(3), object(1)

memory usage: 201.9+ KB

**************************************************


Index(['review_id', 'book_id', 'username', 'text'], dtype='object')

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

review_id      book_id
count  2793.000000  2793.000000
mean   1397.000000   504.693161
std     806.413976   288.472931
min       1.000000     1.000000
25%     699.000000   259.000000
50%    1397.000000   505.000000
75%    2095.000000   753.000000
max    2793.000000  1000.000000

**************************************************

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 2793 entries, 0 to 2792

Data columns (total 4 columns):

 #   Column     Non-Null Count  Dtype 

---  ------     --------------  ----- 

 0   review_id  2793 non-null   int64 

 1   book_id    2793 non-null   int64 

 2   username   2793 non-null   object

 3   text       2793 non-null   object

dtypes: int64(2), object(2)

memory usage: 87.4+ KB

**************************************************


In [10]:
for data in [books, authors, publishers, ratings, reviews]:
    print(data.duplicated().sum())

0

0

0

0

0


В результате ознакомления с данными, можно сделать следующие выводы:

- Названия столбцов оформлены корректно;
- Практически все столбцы приведены к правильному типу данных (исключение - даты);
- В данных отсутствуют пропущенные значения;
- В данных отсутствуют явные дубликты.

## SQL-запросы

Ответим на интересующие нас вопросы.

### Сколько книг вышло после 1 января 2000 года

In [11]:
query = (
'''
    select count(book_id) 
    from books
    where publication_date > '2000-01-01'
        
''')

pd.io.sql.read_sql(query, con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг.

### Количество обзоров и средняя оценка для каждой книги

In [12]:
# отдельно проверим количество обзоров
query = (
'''
    select book_id, count(review_id)
    from reviews
    group by book_id 
    order by count(review_id) desc 
    
''')

pd.io.sql.read_sql(query, con = engine)

book_id  count
0        948      7
1        656      6
2        497      6
3        963      6
4        696      6
..       ...    ...
989      599      1
990      187      1
991      156      1
992      984      1
993      760      1

[994 rows x 2 columns]

In [13]:
# отдельно проверим среднюю оценка
query = (
'''
    select book_id, avg(rating)
    from ratings
    group by book_id
    order by avg(rating) desc 
''')

pd.io.sql.read_sql(query, con = engine)

book_id   avg
0        518  5.00
1        732  5.00
2        347  5.00
3        610  5.00
4        330  5.00
..       ...   ...
995      915  2.25
996      371  2.00
997      316  2.00
998      202  2.00
999      303  1.50

[1000 rows x 2 columns]

In [14]:
# соединим все вместе
query = (
'''
with 

rw as (
            select book_id, count(review_id) as count
            from reviews
            group by book_id 
            
            ),
            
rt as (
            select book_id, avg(rating) as avg
            from ratings
            group by book_id
            
            )
            
select b.book_id, b.title, rw.count, rt.avg
from books as b
LEFT JOIN rw ON b.book_id=rw.book_id
LEFT JOIN rt ON b.book_id=rt.book_id
order by b.book_id


''')

pd.io.sql.read_sql(query, con = engine)

book_id                                              title  count  \
0          1                                       'Salem's Lot    2.0   
1          2                 1 000 Places to See Before You Die    1.0   
2          3  13 Little Blue Envelopes (Little Blue Envelope...    3.0   
3          4  1491: New Revelations of the Americas Before C...    2.0   
4          5                                               1776    4.0   
..       ...                                                ...    ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)    3.0   
996      997                        Xenocide (Ender's Saga  #3)    3.0   
997      998                                    Year of Wonders    4.0   
998      999                        You Suck (A Love Story  #2)    2.0   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...    4.0   

          avg  
0    3.666667  
1    2.500000  
2    4.666667  
3    4.500000  
4    4.000000  
..        ...  
995  3.666667  
996  3.400000  
997  3.200000  
998  4.500000  
999  3.833333  

[1000 rows x 4 columns]

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [16]:
query = (
'''
    select p.publisher_id, p.publisher, count(b.book_id)
    from books as b
    left join publishers as p on b.publisher_id = p.publisher_id
    where b.num_pages > 50
    group by p.publisher_id 
    order by count(b.book_id) desc
    limit 1
''')

pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher  count
0           212  Penguin Books     42

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books.

### Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [17]:
# 
query = (
'''
with 
            
rt as (
            select book_id, count(rating_id),avg(rating) as avg
            from ratings
            group by book_id
            having count(rating_id) >= 50
            )
            
select a.author_id, a.author, avg(rt.avg)
from rt
LEFT JOIN books as b ON rt.book_id=b.book_id
LEFT JOIN authors as a ON b.author_id=a.author_id
group by a.author_id
order by avg(rt.avg) desc
limit 1

''')

pd.io.sql.read_sql(query, con = engine)

author_id                      author       avg
0        236  J.K. Rowling/Mary GrandPré  4.283844

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [20]:
# 
query = (
'''
with 

rw as (
            select username, count(review_id) as count
            from reviews
            group by username
            )
            
select floor(avg(rw.count))
from rw
where rw.username in (
                        select username
                        from ratings
                        group by username
                        having count(rating_id) >= 50
                        )


''')

pd.io.sql.read_sql(query, con = engine)

floor
0   24.0